### PR TITLE
new dataformat for MC weights, slimmed wrt the LHEEventProduct one

### PR DESCRIPTION
**Pull request by govoni**
Tested compiling and pushed in CMSSW_8_0_X_2015-10-25-2300.
Tested running in CMSSW_7_1_14 with available samples.
Might need a review of  SimDataFormats/GeneratorProducts/src/classes_def.xml, as I simply followed the instructions I got from scram b to let it compile.

### CODE REVIEW

**Comment by covarell** [comment]
Questions:
- In  SimDataFormats/GeneratorProducts/src/classes_def.xml we ga back one class version for HepMC::WeightContainer (is this a problem? @bendavid )
- What is the reduction factor in the event size using Lite instead of normal?

**Comment by govoni** [comment]
Ciao Roberto,
an unexpected answer about the size:

[lxplus0060:src] :~> edmEventSize -v testSlimmedWeights.root  | grep LHEEventProduct
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
LHEEventProductLite_addSlimmedWeights__SlimmedWeights. 13397.5 12932.2
LHEEventProduct_externalLHEProducer__LHE. 2782.75 1341

it seems that the Lite version is heavier... I am looking into that.

File: SimDataFormats/GeneratorProducts/interface/LHEEventProductLite.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        hepeup_         = lep->hepeup () ;
+        originalXWGTUP_ = lep->originalXWGTUP () ;
+        npLO_           = lep->npLO () ;
+        npNLO_          = lep->npNLO () ;
+
+        if (lep->pdf ())
+          {
+            this->setPDF (*lep->pdf ()) ;
+          }
+        pdf_.reset(new PDF()) ;
```
Comment (gpetruc) [inline SimDataFormats/GeneratorProducts/interface/LHEEventProductLite.h]: why do you first setPDF and then immediately afterwards reset it to something dummy?

Comment (govoni) [inline SimDataFormats/GeneratorProducts/interface/LHEEventProductLite.h]: it's a bug, and probably the reason of the large difference in size.

**Comment by govoni** [comment]
@covarell a new test, after the bugfix, shows a reduction factor of O(1/2) for 142 events.
The header is still there, so it might be a larger factor in the end.

[lxplus0036:src] :~> edmEventSize -v testSlimmedWeights.root | grep LHEEventProduct
LHEEventProductLite_addSlimmedWeights__SlimmedWeights. 1337.12 777.655
LHEEventProduct_externalLHEProducer__LHE. 2179.28 595.894

**Comment by covarell** [comment]
+1

**Comment by covarell** [comment]
please test

**Comment by covarell** [comment]
@gpetruc , when this is accepted it will require a change in the miniAOD event content definition on your side

**Comment by covarell** [comment]
@govoni @bendavid, what would be the best solution for scheduling the running of the new producer? Should it be done in the "LHE" process or only in the "MINIAOD" process (so keeping the heavier one up to AOD)?

**Comment by govoni** [comment]
My opinion is that having the heavier one in the AOD is important, since the comments field might carry useful information. This is also why I put the conversion from the heavy to the lite format in the lite constructor and not in the producer.
Of course we can think of changing doubles into floats in the heavy version of the data format.

**Comment by bendavid** [comment]
Ah the comments are actually needed in MINIAOD.

They are used in SUSY to keep track of the different model points.

I would say this producer should run in the MINIAOD (ie PAT) step anyways.

File: SimDataFormats/GeneratorProducts/src/classes_def.xml [hunk 2]
```diff
@@ -206,14 +203,19 @@
 	<class name="edm::Wrapper&lt;LHERunInfoProduct&gt;"/>
 	<class name="edm::Wrapper&lt;LHEEventProduct&gt;"/>
 
- <!--needed for backward compatibility between HepMC 2.06.xx and 2.05.yy-->
- <ioread sourceClass = "HepMC::Polarization" version="[1-10]" targetClass="HepMC::Polarization" source = "" target="m_defined">
+<class name="LHEEventProductLite" ClassVersion="-1">
+    <version ClassVersion="-1" checksum="1637646414"/>
```
Comment (davidlange6) [inline SimDataFormats/GeneratorProducts/src/classes_def.xml]: The class version needs to be a number 2 or greater.

Comment (govoni) [inline SimDataFormats/GeneratorProducts/src/classes_def.xml]: where should I push the modification to?
here below my issue.

pietro



scram project CMSSW CMSSW_8_0_X_2015-10-29-2300
cd CMSSW_8_0_X_2015-10-29-2300/src/
cmsenv
git cms-merge-topic govoni:MCWeightsSlimmerSecondTry
[...modifications & compile tests]
git add SimDataFormats/GeneratorProducts/src/classes_def.xml
git commit -m "changing class version number"
git push
fatal: No configured push destination.
Either specify the URL from the command-line or configure a remote repository using

    git remote add <name> <url>

and then push using the remote name

    git push <name>

**Comment by covarell** [comment]
In case you want to update a PR, not open a new one, you should start from the CMSSW area/branch where you did the first commit.

**Comment by covarell** [comment]
+1

**Comment by davidlange6** [comment]
two issues I didn't notice before - please give the class member data variables such as tmp, line (and similar..) a more obvious name - can be tmp_ or line_ to make it clear these are member data. Since you are already keeping tmp as a member data, you should also add the stringstream, which will make these functions much faster.

**Comment by davidlange6** [comment]
ping @govoni

**Comment by govoni** [comment]
- names of the class members for the iterator class changed
 - the stringstream is left as local variable, as the iterator is meant for output as far as I understand, so there's no need for speed

**Comment by covarell** [comment]
Is this issue fixed? 


"Ah the comments are actually needed in MINIAOD.

They are used in SUSY to keep track of the different model points.

I would say this producer should run in the MINIAOD (ie PAT) step anyways."

**Comment by covarell** [comment]
@govoni ping

**Comment by davidlange6** [comment]
@govoni - should we close this PR?

**Comment by govoni** [comment]
The development is not finished.
For me, closing this one and open a new one when manpower is available, or keeping it open does not change much, please decide depending on your usual policies :)

**Comment by davidlange6** [comment]
I'll close it. PRs should be regarded as ready to merge.

**Pull request closed without merging**